### PR TITLE
Newnewfeat/pca

### LINKED ISSUE #158: [Driver] - Make PCA95 driver platform agnostic
### Current Features

this driver depends on stm32 HAL

### Desired Additional Features

We want these to work with all potential hardware. This involves replacing HAL specific code with generalized code that utilized function pointers to use HAL functionality. The way in which we want to handle this is outlined [in this doc](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/430997517/C+Coding+Standards) under the "Driver Generalization" header. 

Also a part of this ticket, find all references to this driver in Shepherd and Cerberus, and fix them to accommodate the new driver structure. A good way to check this is to update your embedded-base submodule to reflect the changes made to this driver, and then try building shepherd and cerberus. If they relied on this device, they will likely fail to build unless you make the appropriate changes

### Screenshots (as needed)

_No response_

### PR DESCRIPTION
**Pull request by rubennoro**
## Changes

Added write and read function pointers, and made i2c_handler of unspecified void* pointer so that it may point to enum that is used in HAL. Experiencing issues in the pdu.c that doesnt reflect these changes, but outside of that it works.

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #158 (issue #)

### CODE REVIEW

File: general/include/pca9539.h [hunk 1]
```diff
@@ -36,31 +35,56 @@ PCA 9539 16 bit GPIO expander.  Datasheet:
...
+
+int pca9539_read_pin(pca9539_t *pca, uint8_t reg_type, uint8_t pin,
+		     uint8_t *buf);
+
+int pca9539_write_reg(pca9539_t *pca, uint8_t reg_type, uint8_t buf);
+
+int pca9539_write_pin(pca9539_t *pca, uint8_t reg_type, uint8_t pin,
+		      uint8_t buf);
+
+/*IGNORE THIS CODE - SERVES AS A REFERENCE
```
Comment (Sabramz) [inline general/include/pca9539.h:72]: Remove this code if its unused

File: general/include/pca9539.h [hunk 2]
```diff
@@ -36,31 +35,56 @@ PCA 9539 16 bit GPIO expander.  Datasheet:
...
+
+	WritePtr write;
+	ReadPtr read;
+
 	uint16_t dev_addr;
 } pca9539_t;
 
-/// Init PCA9539, a 16 bit I2C GPIO expander
-void pca9539_init(pca9539_t *pca, I2C_HandleTypeDef *i2c_handle,
+void pca9539_init(pca9539_t *pca, WritePtr writeFunc, ReadPtr readFunc,
```
Comment (Sabramz) [inline general/include/pca9539.h:59]: Doc comments are missing. Use doxygen, javadoc style comments seen in other repos.

File: general/src/pca9539.c [hunk 3]
```diff
@@ -4,30 +4,112 @@
 
 #define REG_SIZE_BITS 8
 
-HAL_StatusTypeDef pca_write_reg(pca9539_t *pca, uint16_t address, uint8_t *data)
+//RETURNS THE WRITE POINTER INSTEAD
```
Comment (Sabramz) [inline general/src/pca9539.c:7]: This comment is confusing

**Review by Sabramz — CHANGES_REQUESTED** [review]

File: general/src/pca9539.c [hunk 4]
```diff
@@ -34,32 +27,17 @@ int pca_read_reg(pca9539_t *pca, uint16_t address, uint8_t *data)
...
 }
 
-/*IGNORE THIS CODE - LEFT AS A REFERENCE
-void pca9539_init(pca9539_t* pca, I2C_HandleTypeDef* i2c_handle, uint8_t dev_addr)
-{
-	pca->i2c_handle = i2c_handle;
-	pca->dev_addr	= dev_addr << 1u;  shifted one to the left cuz STM says so 
-}
-*/
+//Read PCA9539 Register
```
Comment (Sabramz) [inline general/src/pca9539.c:40]: These comments aren't needed since they are already in the @brief section in the header file.

File: general/include/pca9539.h [hunk 5]
```diff
@@ -58,33 +51,30 @@ typedef struct {
 
 void pca9539_init(pca9539_t *pca, WritePtr writeFunc, ReadPtr readFunc,
 		  uint8_t dev_addr);
+/**
+ * @brief Initialize the PCA9539 Driver
+ */
```
Comment (Sabramz) [inline general/include/pca9539.h:56]: These also need to include definitions for parameters and return types. If you use VScode, you should get the doxygen extension. It autofills all that stuff for you.

**Review by Sabramz — CHANGES_REQUESTED** [review]

File: general/include/pca9539.h [hunk 6]
```diff
@@ -49,32 +58,60 @@ typedef struct {
...
 
-void pca9539_init(pca9539_t *pca, WritePtr writeFunc, ReadPtr readFunc,
-		  uint8_t dev_addr);
 /**
  * @brief Initialize the PCA9539 Driver
+ * 
+ * @param pca 
+ * @param writeFunc 
+ * @param readFunc 
+ * @param dev_addr
```
Comment (Sabramz) [inline general/include/pca9539.h:67]: You should write what each of these mean, even if its completely obvious. Follow how doxygen comments are done in other codebases, like Cerb.

**Review by Sabramz — CHANGES_REQUESTED** [review]

**Review by Sabramz — APPROVED** [review]
